harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,347 @@
+    """
+    if isinstance(base_ds_config, str):
+        base_ds_config = json.load(
+            open(base_ds_config, "r"),
+            object_pairs_hook=config_utils.dict_raise_error_on_duplicate_keys,
+        )
+    else:
+        if not isinstance(base_ds_config, dict):
+            raise TypeError("Expected string or dict for base_ds_config argument.")
+
+    return merge_dicts(cast(Dict[str, Any], base_ds_config), source_ds_dict)
+
+
+class DeepSpeedTrialContext(det.TrialContext, pytorch._PyTorchReducerContext):
+    """Contains runtime information for any Determined workflow that uses the ``DeepSpeedTrial`` API.
+
+    With this class, users can do the following things:
+
+    1. Wrap DeepSpeed model engines which contain the model, optimizer, lr_scheduler, etc.
+       This will make sure Determined can automatically provide gradient aggregation,
+       checkpointing and fault tolerance.  In contrast to :class:`determined.pytorch.PyTorchTrial`,
+       the user does not need to wrap optimizer and lr_scheduler as that should all be instead
+       passed to the DeepSpeed initialize function (see
+       https://www.deepspeed.ai/getting-started/#writing-deepspeed-models) when building the
+       model engine.
+    2. Overwrite a deepspeed config file or dictionary with values from Determined's
+       experiment config to ensure consistency in batch size and support hyperparameter tuning.
+    3. Wrap a custom model parallel configure that should subclass
+       :class:`determined.pytorch.deepspeed.ModelParallelUnit`.  We will automatically set mpu for
+       data parallel and standard pipeline parallel training.  This should only needed if there
+       is additional model parallelism outside of DeepSpeed's supported methods.
+    4. Functionalities inherited from :class:`determined.TrialContext`, including getting
+       the runtime information and properly handling training data in distributed training.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        det.TrialContext.__init__(self, *args, **kwargs)
+        pytorch._PyTorchReducerContext.__init__(self, self.distributed._zmq_allgather)
+
+        self._init_device()
+
+        # Track which types we have issued warnings for in to_device().
+        self._to_device_warned_types = set()  # type: Set[Type]
+
+        # DeepSpeed supports mixed precision through NVidia Apex AMP.  ZeRO optimizer requires
+        # Apex AMP and cannot be used with more complex AMP modes.
+        apex_available = importlib.util.find_spec("apex") is not None
+        if not apex_available:
+            logging.warning(
+                "Missing package APEX is required for ZeRO optimizer support through DeepSpeed."
+            )
+
+        # The following attributes are initialized during the lifetime of
+        # a DeepSpeedTrialContext.
+        self.models = []  # type: List[nn.Module]
+        self.profiler = None  # type: Any
```

Comment:
Why are these publicly named attributes?

Reply:
Update: self.profiler seems like it shouldn't even exist on this class anymore.

Reply:
Agree we should remove self.profiler.  

I think it's okay to keep `models` publich since `context.models` is used in the default `save` and `load` methods for the trial.